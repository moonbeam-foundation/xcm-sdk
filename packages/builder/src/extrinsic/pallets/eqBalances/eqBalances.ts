import { ExtrinsicConfig } from '../../../types/substrate/ExtrinsicConfig';
import type { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import { getExtrinsicAccount } from '../../ExtrinsicBuilder.utils';

export enum EqBalancesFee {
  SovereignAccWillPay = 'SovereignAccWillPay',
  TargetChainAccWillPay = 'TargetChainAccWillPay',
  ThisAccWillPay = 'ThisAccWillPay',
}

const pallet = 'eqBalances';

export function eqBalances() {
  return {
    xcmTransfer: (): ExtrinsicConfigBuilder => ({
      build: ({ destinationAddress, asset, destination }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'xcmTransfer',
          getArgs: () => [
            asset.getAssetId(),
            asset.amount,
            {
              parents: 1,
              interior: {
                X2: [
                  {
                    Parachain: destination.parachainId,
                  },
                  getExtrinsicAccount(destinationAddress),
                ],
              },
            },
            EqBalancesFee.ThisAccWillPay,
          ],
        }),
    }),
    transferXcm: (): ExtrinsicConfigBuilder => ({
      build: ({ destinationAddress: address, asset, destination, fee }) =>
        new ExtrinsicConfig({
          module: pallet,
          func: 'transferXcm',
          getArgs: () => {
            const amountWithoutFee =
              asset.amount - fee.amount > 0n ? asset.amount - fee.amount : 0n;

            return [
              [
                asset.getAssetId(),
                asset.isSame(fee) ? amountWithoutFee : asset.amount,
              ],
              [fee.getAssetId(), fee.amount],
              {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: destination.parachainId,
                    },
                    getExtrinsicAccount(address),
                  ],
                },
              },
            ];
          },
        }),
    }),
  };
}
