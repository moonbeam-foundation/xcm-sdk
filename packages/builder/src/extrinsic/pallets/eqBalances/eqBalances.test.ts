import { describe, expect, it } from 'vitest';

import {
  buildParachainParamsMock,
  buildParamsMock,
} from '../../../../fixtures';
import { eqBalances } from './eqBalances';

describe('eqBalances', () => {
  describe('xcmTransfer', () => {
    const extrinsic = eqBalances().xcmTransfer().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });

  describe('transferXcm', () => {
    const extrinsic = eqBalances().transferXcm().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });

  describe('xcmTransfer to parachain', () => {
    const extrinsic = eqBalances()
      .xcmTransfer()
      .build(buildParachainParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });

  describe('transferXcm to parachain', () => {
    const extrinsic = eqBalances()
      .transferXcm()
      .build(buildParachainParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });
});
