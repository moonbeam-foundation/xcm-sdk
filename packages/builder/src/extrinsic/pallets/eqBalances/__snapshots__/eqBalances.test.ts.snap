// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eqBalances > transferXcm > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferXcm",
  "getArgs": [Function],
  "module": "eqBalances",
}
`;

exports[`eqBalances > transferXcm > should get correct arguments 1`] = `
[
  [
    "USDT",
    99000000000n,
  ],
  [
    "RMRK",
    5000000000n,
  ],
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
]
`;

exports[`eqBalances > transferXcm to parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferXcm",
  "getArgs": [Function],
  "module": "eqBalances",
}
`;

exports[`eqBalances > transferXcm to parachain > should get correct arguments 1`] = `
[
  [
    "USDT",
    99000000000n,
  ],
  [
    "RMRK",
    5000000000n,
  ],
  {
    "interior": {
      "X2": [
        {
          "Parachain": 2032,
        },
        {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      ],
    },
    "parents": 1,
  },
]
`;

exports[`eqBalances > xcmTransfer > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "xcmTransfer",
  "getArgs": [Function],
  "module": "eqBalances",
}
`;

exports[`eqBalances > xcmTransfer > should get correct arguments 1`] = `
[
  "USDT",
  99000000000n,
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  "ThisAccWillPay",
]
`;

exports[`eqBalances > xcmTransfer to parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "xcmTransfer",
  "getArgs": [Function],
  "module": "eqBalances",
}
`;

exports[`eqBalances > xcmTransfer to parachain > should get correct arguments 1`] = `
[
  "USDT",
  99000000000n,
  {
    "interior": {
      "X2": [
        {
          "Parachain": 2032,
        },
        {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      ],
    },
    "parents": 1,
  },
  "ThisAccWillPay",
]
`;
